src/oidc/tokenRefresher.ts

```diff
@@ -0,0 +1,135 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { IdTokenClaims, OidcClient, WebStorageStateStore } from "oidc-client-ts";
+
+import { TokenRefreshFunction } from "../http-api";
+import { IDelegatedAuthConfig } from "../client";
+import { generateScope } from "./authorize";
+import { discoverAndValidateAuthenticationConfig } from "./discovery";
+import { logger } from "../logger";
+
+/**
+ * @experimental
+ * Class responsible to refreshing OIDC access tokens
+ */
+export abstract class OidcTokenRefresher {
+    public readonly oidcClientReady!: Promise<void>;
+    private oidcClient!: OidcClient;
+    private inflightRefreshRequest?: ReturnType<TokenRefreshFunction>;
+
+    public constructor(
+        /**
+         * Delegated auth config as found in matrix client .well-known
+         */
+        authConfig: IDelegatedAuthConfig,
+        /**
+         * id of this client as registered with the OP
+         */
+        clientId: string,
+        /**
+         * redirectUri as registered with OP
+         */
+        redirectUri: string,
+        deviceId: string,
+        /**
+         * idTokenClaims as returned from authorization grant
+         * used to validate tokens
+         */
+        private readonly idTokenClaims: IdTokenClaims,
+    ) {
+        this.oidcClientReady = this.initialiseOidcClient(authConfig, clientId, deviceId, redirectUri);
+    }
+
+    private async initialiseOidcClient(
+        authConfig: IDelegatedAuthConfig,
+        clientId: string,
+        deviceId: string,
+        redirectUri: string,
+    ): Promise<void> {
+        const config = await discoverAndValidateAuthenticationConfig(authConfig);
+
+        const scope = generateScope(deviceId);
+
+        this.oidcClient = new OidcClient({
+            ...config.metadata,
+            client_id: clientId,
+            scope,
+            redirect_uri: redirectUri,
+            authority: config.metadata.issuer,
+            stateStore: new WebStorageStateStore({ prefix: "mx_oidc_", store: window.sessionStorage }),
+        });
+    }
+
+    /**
+     * Attempt token refresh using given access token
+     * When successful, persist new tokens
+     * @param refreshToken - refresh token to use in request with token issuer
+     * @returns tokens - Promise that resolves with new access and refresh tokens
+     * @throws when token refresh fails
+     */
+    public async doRefreshAccessToken(refreshToken: string): ReturnType<TokenRefreshFunction> {
+        if (!this.inflightRefreshRequest) {
+            this.inflightRefreshRequest = this.getNewToken(refreshToken);
+        }
+        try {
+            const tokens = await this.inflightRefreshRequest;
+            await this.persistTokens(tokens);
```

Comment:
should this be in `getNewToken`, so as to avoid needlessly persisting the same result multiple times?